<img src="img/dsci513_header2.png" width="600">

# Lecture 6: Views, CTEs, window functions, indexing

## Lecture outline

- Views
- Common Table Expressions (CTEs)
- Window functions
- indexing and performance

<br><br><br>

## Views

In the last lecture, we tried to write a query for the following example:

---

**Example:** Which countries speak at least one language that is not spoken an any other country in their continent?

---

<br><br><br>

This query involves data both from the `country` table as well and the `countrylanguage` table.

On the other hand, we also had to use an `EXISTS` subquery. This was why we first created an intermediate query to first store the results from joining the two tables, and then add the subquery part.

While that solution worked, it might not always be ideal to store the data in another table (why?).

<br><br><br>

Although the separation of country data from language data makes sense from a logical point of view, it might happen that country + language data are the ones that are accessed the most by our users. Also it might be the case that our users only need names, continents, and population data from the `country` table, and language and percentage from the `countrylanguage` table. In this situation, we can create a **view** of the data that we have in our database.

A view is a **named** result of a `SELECT` query that behaves just like a table; i.e. you can query it just like you query any other table in your database.

<br><br><br>

### How to create and delete a view

```sql
CREATE VIEW view_name AS
    select_statment
;
```

<br>

```sql
DROP VIEW [IF EXISTS] view_name;
```

<br><br><br>

### Properties of a view

- A view acts as if it is a table, while it's actually only a way of looking at the data differently. This is why they're sometimes called **virtual tables**. As opposed to the temporary table approach, there is no duplication of data.

- A view is always **current and dynamic**. Whenever we access the view, the query that generates it is re-run and we will see an up-to-date version of the data. This may not be true with a temporary table.

- **Views persist.** You might argue that a temporary table only occupies space in the duration of a session, but it's not certain if that's truly an advantage. While it's good that the taken space is released after closing the connection, each time a user needs that particular query, the table also needs to be recreated.

- Views can **hide unnecessary details, or add details that are computed on demand**. Especially in complex databases, not every database user needs to, or wants to have access to all details of the dataset. Can you think of an example?

- With views we can also **manage what different users can see**. For example, in the database of a company not every department needs (or should) have access to all financial information of employees. But they may need their account and SIN numbers to be able to pay them for the job the do for that department. This can be easily managed with views.

- Views do not support constraints.

<br><br><br>

### Materialized views: the best of both worlds

Sometimes the query underlying a view might be too slow to compute on the fly in a simple view. In this situation, there is a way to physically store a view. In other words, views can be **materialized**.

A materialized view stores the results of the view query on the physical disk, instead of recomputing the entire query, which is why materialized views are much faster to use.

<br><br><br>

Here is the syntax for creating and dropping materialized views:

```sql
CREATE MATERIALIZED VIEW my_mat_view AS
    select_statement
;

DROP MATERIALIZED VIEW my_mat_view;
```

<br><br><br>

Like almost anything in this world, you won't gain anything unless you lose something else. Same is true with materialized views. The **trade-offs** you need to be aware of when using materialized views:

- Just like a temporary table, a materialized view occupies disk space

- The data in a materialized view needs to be refreshed if the data in the related tables are modified.

<br><br><br>

Materialized views can be refreshed using the following syntax:

```sql
REFRESH MATERIALIZED VIEW [CONCURRENTLY] my_mat_view;
```

**Note:** If the keyword `CONCURRENTLY` is used, the view will be usable during the refresh process, but there's no guarantee that the data will be up to date.

<br><br><br>

**Takeaway:** Using a materialized view is justified if the query takes **a lot of time but little space**. A simple query, on the other hand, is best suited to a query that takes **a considerable amount of space, but relatively little time**.

<br><br><br>

---

**Question:** Can you think of two advantages of a temporary table over a simple view?

---

<br><br><br>

In [1]:
%load_ext sql
%config SqlMagic.displaylimit = 30

In [2]:
import json
import urllib.parse

with open('data/credentials.json') as f:
    login = json.load(f)
    
username = login['user']
password = urllib.parse.quote(login['password'])
host = login['host']
port = login['port']

In [3]:
%sql postgresql://{username}:{password}@{host}/imdb

'Connected: postgres@imdb'

In this example here, I have intentionally written a slow query (involving a correlated subquery) to find movies whose number of votes is higher than the average number of votes of all movies belonging to the same genre. Running this query will take some time:

In [4]:
%%sql

SELECT
    *
FROM
    movies m
JOIN
    movie_genres mg
ON
    m.id = mg.movie_id
WHERE        
    rating > 8
    AND
    nvotes > (
        SELECT
            AVG(m2.nvotes)
        FROM
            movies m2
        JOIN
            movie_genres mg2
        ON
            m2.id = mg2.movie_id
        WHERE
            mg.genre = mg2.genre
            AND
            m.id <> m2.id
    )
;

 * postgresql://postgres:***@localhost/imdb
795 rows affected.


[(10042876, 'Rashomon', 'Rashômon', 1950, None, 88, 8.2, 138304, 10042876, 'crime'),
 (10042876, 'Rashomon', 'Rashômon', 1950, None, 88, 8.2, 138304, 10042876, 'drama'),
 (10042876, 'Rashomon', 'Rashômon', 1950, None, 88, 8.2, 138304, 10042876, 'mystery'),
 (10043014, 'Sunset Blvd.', None, 1950, None, 110, 8.4, 183282, 10043014, 'drama'),
 (10043014, 'Sunset Blvd.', None, 1950, None, 110, 8.4, 183282, 10043014, 'film-noir'),
 (10044741, 'Ikiru', None, 1952, None, 143, 8.3, 60061, 10044741, 'drama'),
 (10044837, 'Limelight', None, 1952, None, 137, 8.1, 16445, 10044837, 'music'),
 (10045152, "Singin' in the Rain", None, 1952, None, 103, 8.3, 201077, 10045152, 'comedy'),
 (10045152, "Singin' in the Rain", None, 1952, None, 103, 8.3, 201077, 10045152, 'musical'),
 (10045152, "Singin' in the Rain", None, 1952, None, 103, 8.3, 201077, 10045152, 'romance'),
 (10046268, 'The Wages of Fear', 'Le salaire de la peur', 1953, None, 131, 8.1, 49244, 10046268, 'drama'),
 (10046268, 'The Wages of Fear', 'Le salaire de la peur', 1953, None, 131, 8.1, 49244, 10046268, 'thriller'),
 (10046478, 'Ugetsu', 'Ugetsu monogatari', 1953, None, 96, 8.3, 18356, 10046478, 'war'),
 (10046912, 'Dial M for Murder', None, 1954, None, 105, 8.2, 143268, 10046912, 'crime'),
 (10046912, 'Dial M for Murder', None, 1954, None, 105, 8.2, 143268, 10046912, 'thriller'),
 (10047296, 'On the Waterfront', None, 1954, None, 108, 8.1, 130916, 10047296, 'crime'),
 (10047296, 'On the Waterfront', None, 1954, None, 108, 8.1, 130916, 10047296, 'drama'),
 (10047296, 'On the Waterfront', None, 1954, None, 108, 8.1, 130916, 10047296, 'thriller'),
 (10047396, 'Rear Window', None, 1954, None, 112, 8.5, 407290, 10047396, 'mystery'),
 (10047396, 'Rear Window', None, 1954, None, 112, 8.5, 407290, 10047396, 'thriller'),
 (10050083, '12 Angry Men', None, 1957, None, 96, 8.9, 610724, 10050083, 'drama'),
 (10050613, 'Throne of Blood', 'Kumonosu-jô', 1957, None, 110, 8.1, 40639, 10050613, 'drama'),
 (10050613, 'Throne of Blood', 'Kumonosu-jô', 1957, None, 110, 8.1, 40639, 10050613, 'history'),
 (10050783, 'The Nights of Cabiria', 'Le notti di Cabiria', 1957, None, 110, 8.1, 39350, 10050783, 'drama'),
 (10050825, 'Paths of Glory', None, 1957, None, 88, 8.4, 160459, 10050825, 'drama'),
 (10050825, 'Paths of Glory', None, 1957, None, 88, 8.4, 160459, 10050825, 'war'),
 (10050986, 'Wild Strawberries', 'Smultronstället', 1957, None, 91, 8.2, 85988, 10050986, 'drama'),
 (10050986, 'Wild Strawberries', 'Smultronstället', 1957, None, 91, 8.2, 85988, 10050986, 'romance'),
 (10051036, 'Sweet Smell of Success', None, 1957, None, 96, 8.1, 24897, 10051036, 'drama'),
 (10051036, 'Sweet Smell of Success', None, 1957, None, 96, 8.1, 24897, 10051036, 'film-noir'),
 (10050976, 'The Seventh Seal', 'Det sjunde inseglet', 1957, None, 96, 8.2, 146553, 10050976, 'drama'),
 (10050976, 'The Seventh Seal', 'Det sjunde inseglet', 1957, None, 96, 8.2, 146553, 10050976, 'fantasy'),
 (10051201, 'Witness for the Prosecution', None, 1957, None, 116, 8.4, 95259, 10051201, 'crime'),
 (10051201, 'Witness for the Prosecution', None, 1957, None, 116, 8.4, 95259, 10051201, 'drama'),
 (10051201, 'Witness for the Prosecution', None, 1957, None, 116, 8.4, 95259, 10051201, 'film-noir'),
 (10051808, 'The Hidden Fortress', 'Kakushi-toride no san-akunin', 1958, None, 139, 8.1, 30431, 10051808, 'drama'),
 (10052357, 'Vertigo', None, 1958, None, 128, 8.3, 327884, 10052357, 'mystery'),
 (10052357, 'Vertigo', None, 1958, None, 128, 8.3, 327884, 10052357, 'romance'),
 (10052357, 'Vertigo', None, 1958, None, 128, 8.3, 327884, 10052357, 'thriller'),
 (10052618, 'Ben-Hur', None, 1959, None, 212, 8.1, 204418, 10052618, 'adventure'),
 (10052618, 'Ben-Hur', None, 1959, None, 212, 8.1, 204418, 10052618, 'drama'),
 (10052618, 'Ben-Hur', None, 1959, None, 212, 8.1, 204418, 10052618, 'history'),
 (10053291, 'Some Like It Hot', None, 1959, None, 121, 8.2, 224857, 10053291, 'comedy'),
 (10053291, 'Some Like It Hot', None, 1959, None, 121, 8.2, 224857,

Let's create a simple view for this query:

In [5]:
%%sql

CREATE VIEW
    simple_view
AS
    SELECT
        *
    FROM
        movies m
    JOIN
        movie_genres mg
    ON
        m.id = mg.movie_id
    WHERE        
        rating > 8
        AND
        nvotes > (
            SELECT
                AVG(m2.nvotes)
            FROM
                movies m2
            JOIN
                movie_genres mg2
            ON
                m2.id = mg2.movie_id
            WHERE
                mg.genre = mg2.genre
                AND
                m.id <> m2.id
        )
    ;

 * postgresql://postgres:***@localhost/imdb
Done.


[]

Retrieving data from this view will be as slow as running the original query:

In [6]:
%%timeit -r 1 -n 1
%%sql

SELECT * FROM simple_view;

 * postgresql://postgres:***@localhost/imdb
795 rows affected.
13.6 s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


<br><br><br>

The result of this query takes little space (only 795 rows), but it takes a long time to compute, which makes it a good candidate for a materialized view. Note that creating the materialized view will be slow the first time:

In [7]:
%%timeit -r 1 -n 1
%%sql

CREATE MATERIALIZED VIEW
    mat_view
AS
    SELECT
        *
    FROM
        movies m
    JOIN
        movie_genres mg
    ON
        m.id = mg.movie_id
    WHERE        
        rating > 8
        AND
        nvotes > (
            SELECT
                AVG(m2.nvotes)
            FROM
                movies m2
            JOIN
                movie_genres mg2
            ON
                m2.id = mg2.movie_id
            WHERE
                mg.genre = mg2.genre
                AND
                m.id <> m2.id
        )
;

 * postgresql://postgres:***@localhost/imdb
795 rows affected.
13.4 s ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


But retrieving data from the materialized view should be almost instantaneous:

In [8]:
%%timeit -r 1 -n 1
%%sql

SELECT * FROM mat_view;

 * postgresql://postgres:***@localhost/imdb
795 rows affected.
3.55 ms ± 0 ns per loop (mean ± std. dev. of 1 run, 1 loop each)


Let's not forget to clean up:

In [9]:
%%sql

DROP VIEW simple_view;
DROP MATERIALIZED VIEW mat_view;

 * postgresql://postgres:***@localhost/imdb
Done.
Done.


[]

<br><br><br>

## Common Table Expressions (CTEs)

It's not uncommon for our queries to become overly complex. For example, consider the following question:

---

**Example:** Which countries speak at least one language that is not spoken an any other country in their continent?

---

A query to answer the above question involves joining tables first to have both country and language data in the same place, and then use a subquery.

To that end, we can create a temporary table first and then use the subquery on that table:

```sql
DROP TABLE IF EXISTS ccl;

CREATE TEMPORARY TABLE ccl AS (
    SELECT
        co.name, co.continent, cl.language
    FROM
        country co
    JOIN
        countrylanguage cl
    ON
        co.code = cl.countrycode
)
;

SELECT
    t1.*
FROM
    ccl t1
WHERE
    NOT EXISTS (
        SELECT
            *
        FROM
            ccl t2
        WHERE
            t1.name <> t2.name
            AND
            t1.continent = t2.continent
            AND
            t1.language = t2.language
    )
;
```

<br><br><br>

For this situation, we may not want to create a **temporary table** because:

- copying the data of two entire tables might create a disk space problem during a session

- if the data is modified in anyways before another subsequent query, it won't be reflected in the temporary table

- the temporary table has to be created each time a connection to the database is made

<br><br><br>

Creating a **view** looks like a nice alternative because it doesn't pose any of the above problems, but:

- The purpose of a view is to provide a more or less permanent way of accessing the data or hiding details. It doesn't make sense to create a view that is intended just for a very specific query and useless otherwise. For example, do we really want a view in our database that shows countries and languages just in Asia and Europe, or is it just for the purpose of a special query?

- Creating views requires special access privileges, which as a user we might not have!

<br><br><br>

Standard SQL has an elegant way of addressing this issue: **Common Table Expressions**

A common table expression is a temporary intermediate result set that we can use within a `SELECT`, `INSERT`, `UPDATE`, or `DELETE` statement. The syntax for a CTE is:

```sql
WITH
    expression_name [(column_names, ...)]
AS (
    query
)
query
;
```

<br><br><br>

Let's take a look at a few examples:

In [10]:
%sql postgresql://{username}:{password}@{host}:{port}/world

'Connected: postgres@world'

The following CTE creates an intermediate result set from the `country` and `countrylanguage` tables:

In [11]:
%%sql

WITH my_cte AS (
    SELECT
        name, population, language
    FROM
        country co
    JOIN
        countrylanguage cl
    ON
        cl.countrycode = co.code
)
SELECT * FROM my_cte
;

   postgresql://postgres:***@localhost/imdb
 * postgresql://postgres:***@localhost:5432/world
984 rows affected.


[('Afghanistan', 22720000, 'Pashto'),
 ('Netherlands', 15864000, 'Dutch'),
 ('Netherlands Antilles', 217000, 'Papiamento'),
 ('Albania', 3401200, 'Albaniana'),
 ('Algeria', 31471000, 'Arabic'),
 ('American Samoa', 68000, 'Samoan'),
 ('Andorra', 78000, 'Spanish'),
 ('Angola', 12878000, 'Ovimbundu'),
 ('Anguilla', 8000, 'English'),
 ('Antigua and Barbuda', 68000, 'Creole English'),
 ('United Arab Emirates', 2441000, 'Arabic'),
 ('Argentina', 37032000, 'Spanish'),
 ('Armenia', 3520000, 'Armenian'),
 ('Aruba', 103000, 'Papiamento'),
 ('Australia', 18886000, 'English'),
 ('Azerbaijan', 7734000, 'Azerbaijani'),
 ('Bahamas', 307000, 'Creole English'),
 ('Bahrain', 617000, 'Arabic'),
 ('Bangladesh', 129155000, 'Bengali'),
 ('Barbados', 270000, 'Bajan'),
 ('Belgium', 10239000, 'Dutch'),
 ('Belize', 241000, 'English'),
 ('Benin', 6097000, 'Fon'),
 ('Bermuda', 65000, 'English'),
 ('Bhutan', 2124000, 'Dzongkha'),
 ('Bolivia', 8329000, 'Spanish'),
 ('Bosnia and Herzegovina', 3972000, 'Serbo-Croatian'),
 ('Botswana', 1622000, 'Tswana'),
 ('Brazil', 170115000, 'Portuguese'),
 ('United Kingdom', 59623400, 'English'),
 ('Virgin Islands, British', 21000, 'English'),
 ('Brunei', 328000, 'Malay'),
 ('Bulgaria', 8190900, 'Bulgariana'),
 ('Burkina Faso', 11937000, 'Mossi'),
 ('Burundi', 6695000, 'Kirundi'),
 ('Cayman Islands', 38000, 'English'),
 ('Chile', 15211000, 'Spanish'),
 ('Cook Islands', 20000, 'Maori'),
 ('Costa Rica', 4023000, 'Spanish'),
 ('Djibouti', 638000, 'Somali'),
 ('Dominica', 71000, 'Creole English'),
 ('Dominican Republic', 8495000, 'Spanish'),
 ('Ecuador', 12646000, 'Spanish'),
 ('Egypt', 68470000, 'Arabic'),
 ('El Salvador', 6276000, 'Spanish'),
 ('Eritrea', 3850000, 'Tigrinja'),
 ('Spain', 39441700, 'Spanish'),
 ('South Africa', 40377000, 'Zulu'),
 ('Ethiopia', 62565000, 'Oromo'),
 ('Falkland Islands', 2000, 'English'),
 ('Fiji Islands', 817000, 'Fijian'),
 ('Philippines', 75967000, 'Pilipino'),
 ('Faroe Islands', 43000, 'Faroese'),
 ('Gabon', 1226000, 'Fang'),
 ('Gambia', 1305000, 'Malinke'),
 ('Georgia', 4968000, 'Georgiana'),
 ('Ghana', 20212000, 'Akan'),
 ('Gibraltar', 25000, 'English'),
 ('Grenada', 94000, 'Creole English'),
 ('Greenland', 56000, 'Greenlandic'),
 ('Guadeloupe', 456000, 'Creole French'),
 ('Guam', 168000, 'English'),
 ('Guatemala', 11385000, 'Spanish'),
 ('Guinea', 7430000, 'Ful'),
 ('Guinea-Bissau', 1213000, 'Crioulo'),
 ('Guyana', 861000, 'Creole English'),
 ('Haiti', 8222000, 'Haiti Creole'),
 ('Honduras', 6485000, 'Spanish'),
 ('Hong Kong', 6782000, 'Canton Chinese'),
 ('Svalbard and Jan Mayen', 3200, 'Norwegian'),
 ('Indonesia', 212107000, 'Javanese'),
 ('India', 1013662000, 'Hindi'),
 ('Iraq', 23115000, 'Arabic'),
 ('Iran', 67702000, 'Persian'),
 ('Ireland', 3775100, 'English'),
 ('Iceland', 279000, 'Icelandic'),
 ('Israel', 6217000, 'Hebrew'),
 ('Italy', 57680000, 'Italian'),
 ('East Timor', 885000, 'Sunda'),
 ('Austria', 8091800, 'German'),
 ('Jamaica', 2583000, 'Creole English'),
 ('Japan', 126714000, 'Japanese'),
 ('Yemen', 18112000, 'Arabic'),
 ('Jordan', 5083000, 'Arabic'),
 ('Christmas Island', 2500, 'Chinese'),
 ('Yugoslavia', 10640000, 'Serbo-Croatian'),
 ('Cambodia', 11168000, 'Khmer'),
 ('Cameroon', 15085000, 'Fang'),
 ('Canada', 31147000, 'English'),
 ('Cape Verde', 428000, 'Crioulo'),
 ('Kazakstan', 16223000, 'Kazakh'),
 ('Kenya', 30080000, 'Kikuyu'),
 ('Central African Republic', 3615000, 'Gbaya'),
 ('China', 1277558000, 'Chinese'),
 ('Kyrgyzstan', 4699000, 'Kirgiz'),
 ('Kiribati', 83000, 'Kiribati'),
 ('Colombia', 42321000, 'Spanish'),
 ('Comoros', 578000, 'Comorian'),
 ('Congo', 2943000, 'Kongo'),
 ('Congo, The Democratic Republic of the', 51654000, 'Luba'),
 ('Cocos (Keeling) Islands', 600, 'Malay'),
 ('North Korea', 24039000, 'Korean'),
 ('South Korea', 46844000, 'Korean'),
 ('Greece', 10545700, 'Greek'),
 ('Croatia', 4473000, 'Serbo-Croatian'),
 ('Cuba', 11201000, 'Spanish'),
 ('Kuwait', 1972000, 'Arabic'),
 ('Cyprus', 754700, 'Greek'),
 ('Laos', 5433000, 'Lao'),
 ('Latvia', 24242

Let's say we want to count the number of official languages in each country. We can write this query as:

In [12]:
%%sql

WITH my_cte AS (
    SELECT
        co.name, cl.language
    FROM
        country co
    JOIN
        countrylanguage cl
    ON
        cl.countrycode = co.code
    WHERE
        cl.isofficial = TRUE
)
SELECT
    name, COUNT(language)
FROM
    my_cte
GROUP BY
    name
ORDER BY
    COUNT(language) DESC
;

   postgresql://postgres:***@localhost/imdb
 * postgresql://postgres:***@localhost:5432/world
190 rows affected.


[('Switzerland', 4),
 ('South Africa', 4),
 ('Singapore', 3),
 ('Peru', 3),
 ('Bolivia', 3),
 ('Vanuatu', 3),
 ('Belgium', 3),
 ('Luxembourg', 3),
 ('Seychelles', 2),
 ('Cyprus', 2),
 ('Netherlands Antilles', 2),
 ('Romania', 2),
 ('Kyrgyzstan', 2),
 ('Burundi', 2),
 ('Afghanistan', 2),
 ('Guam', 2),
 ('Togo', 2),
 ('Israel', 2),
 ('Faroe Islands', 2),
 ('Nauru', 2),
 ('Canada', 2),
 ('Palau', 2),
 ('Tonga', 2),
 ('Madagascar', 2),
 ('Greenland', 2),
 ('Tuvalu', 2),
 ('Samoa', 2),
 ('Malta', 2),
 ('Rwanda', 2),
 ('Sri Lanka', 2),
 ('Somalia', 2),
 ('American Samoa', 2),
 ('Marshall Islands', 2),
 ('Belarus', 2),
 ('Finland', 2),
 ('Lesotho', 2),
 ('Paraguay', 2),
 ('Ireland', 2),
 ('Uruguay', 1),
 ('United Kingdom', 1),
 ('Germany', 1),
 ('Eritrea', 1),
 ('Barbados', 1),
 ('Saint Vincent and the Grenadines', 1),
 ('Cook Islands', 1),
 ('Slovenia', 1),
 ('Azerbaijan', 1),
 ('Greece', 1),
 ('Egypt', 1),
 ('Chad', 1),
 ('Saint Kitts and Nevis', 1),
 ('Chile', 1),
 ('Norfolk Island', 1),
 ('Yugoslavia', 1),
 ('United States Minor Outlying Islands', 1),
 ('Japan', 1),
 ('Macao', 1),
 ('Iraq', 1),
 ('New Zealand', 1),
 ('Ecuador', 1),
 ('Norway', 1),
 ('Pakistan', 1),
 ('Brazil', 1),
 ('Guatemala', 1),
 ('Panama', 1),
 ('Yemen', 1),
 ('Tunisia', 1),
 ('Sudan', 1),
 ('Swaziland', 1),
 ('Morocco', 1),
 ('Christmas Island', 1),
 ('Moldova', 1),
 ('Myanmar', 1),
 ('Mexico', 1),
 ('Poland', 1),
 ('Lebanon', 1),
 ('Turks and Caicos Islands', 1),
 ('Virgin Islands, U.S.', 1),
 ('Indonesia', 1),
 ('Kiribati', 1),
 ('Virgin Islands, British', 1),
 ('Jordan', 1),
 ('Macedonia', 1),
 ('Uzbekistan', 1),
 ('Brunei', 1),
 ('Fiji Islands', 1),
 ('Cayman Islands', 1),
 ('Cuba', 1),
 ('Tokelau', 1),
 ('France', 1),
 ('Djibouti', 1),
 ('Syria', 1),
 ('Senegal', 1),
 ('Iceland', 1),
 ('Hong Kong', 1),
 ('Philippines', 1),
 ('Bangladesh', 1),
 ('Turkey', 1),
 ('Cocos (Keeling) Islands', 1),
 ('Zimbabwe', 1),
 ('Comoros', 1),
 ('Saint Lucia', 1),
 ('Netherlands', 1),
 ('Bahrain', 1),
 ('Spain', 1),
 ('Saint Pierre and Miquelon', 1),
 ('Falkland Islands', 1),
 ('Malawi', 1),
 ('Taiwan', 1),
 ('Bhutan', 1),
 ('Niue', 1),
 ('Italy', 1),
 ('Oman', 1),
 ('South Korea', 1),
 ('Anguilla', 1),
 ('Svalbard and Jan Mayen', 1),
 ('Bermuda', 1),
 ('Argentina', 1),
 ('Liechtenstein', 1),
 ('Puerto Rico', 1),
 ('India', 1),
 ('Iran', 1),
 ('French Polynesia', 1),
 ('Estonia', 1),
 ('Vietnam', 1),
 ('Kazakstan', 1),
 ('Guadeloupe', 1),
 ('Denmark', 1),
 ('Mongolia', 1),
 ('Northern Mariana Islands', 1),
 ('Hungary', 1),
 ('Honduras', 1),
 ('Belize', 1),
 ('Russian Federation', 1),
 ('Austria', 1),
 ('Guinea-Bissau', 1),
 ('Western Sahara', 1),
 ('Lithuania', 1),
 ('Bulgaria', 1),
 ('Turkmenistan', 1),
 ('Croatia', 1),
 ('Libyan Arab Jamahiriya', 1),
 ('Aruba', 1),
 ('East Timor', 1),
 ('Nicaragua', 1),
 ('Nepal', 1),
 ('Tanzania', 1),
 ('Costa Rica', 1),
 ('Haiti', 1),
 ('Gibraltar', 1),
 ('Antigua and Barbuda', 1),
 ('Andorra', 1),
 ('United States', 1),
 ('Mayotte', 1),
 ('Venezuela', 1),
 ('Czech Republic', 1),
 ('Sweden', 1),
 ('Dominican Republic', 1),
 ('Saint Helena', 1),
 ('Cambodia', 1),
 ('San Marino', 1),
 ('Laos', 1),
 ('Portugal', 1),
 ('Colombia', 1),
 ('Albania', 1),
 ('Ukraine', 1),
 ('Saudi Arabia', 1),
 ('Latvia', 1),
 ('North Korea', 1),
 ('Holy See (Vatican City State)', 1),
 ('Algeria', 1),
 ('Cape Verde', 1),
 ('Slovakia', 1),
 ('Maldives', 1),
 ('Malaysia', 1),
 ('Kuwait', 1),
 ('Bosnia and Herzegovina', 1),
 ('Tajikistan', 1),
 ('China', 1),
 ('New Caledonia', 1),
 ('Armenia', 1),
 ('Qatar', 1),
 ('Martinique', 1),
 ('Australia', 1),
 ('United Arab Emirates', 1),
 ('Georgia', 1),
 ('Monaco', 1),
 ('Montserrat', 1),
 ('Thailand', 1),
 ('El Salvador', 1)]

We can also rewrite this query from the last lecture using a CTE:

**Question:** Which countries speak at least one language that is not spoken in any other country on their continent?

In [13]:
%%sql

WITH ccl AS (
    SELECT
        co.name, co.continent, cl.language
    FROM
        country co
    JOIN
        countrylanguage cl
    ON
        co.code = cl.countrycode
)
SELECT
    t1.*
FROM
    ccl t1
WHERE
    NOT EXISTS (
        SELECT
            *
        FROM
            ccl t2
        WHERE
            t1.name <> t2.name
            AND
            t1.continent = t2.continent
            AND
            t1.language = t2.language
    )

   postgresql://postgres:***@localhost/imdb
 * postgresql://postgres:***@localhost:5432/world
362 rows affected.


[('Angola', 'Africa', 'Ovimbundu'),
 ('Barbados', 'North America', 'Bajan'),
 ('Benin', 'Africa', 'Fon'),
 ('Bhutan', 'Asia', 'Dzongkha'),
 ('Burundi', 'Africa', 'Kirundi'),
 ('Ethiopia', 'Africa', 'Oromo'),
 ('Falkland Islands', 'South America', 'English'),
 ('Fiji Islands', 'Oceania', 'Fijian'),
 ('Philippines', 'Asia', 'Pilipino'),
 ('Faroe Islands', 'Europe', 'Faroese'),
 ('Georgia', 'Asia', 'Georgiana'),
 ('Greenland', 'North America', 'Greenlandic'),
 ('Haiti', 'North America', 'Haiti Creole'),
 ('Indonesia', 'Asia', 'Javanese'),
 ('Iceland', 'Europe', 'Icelandic'),
 ('Japan', 'Asia', 'Japanese'),
 ('Kenya', 'Africa', 'Kikuyu'),
 ('Central African Republic', 'Africa', 'Gbaya'),
 ('Kyrgyzstan', 'Asia', 'Kirgiz'),
 ('Congo, The Democratic Republic of the', 'Africa', 'Luba'),
 ('Cocos (Keeling) Islands', 'Oceania', 'Malay'),
 ('Cyprus', 'Asia', 'Greek'),
 ('Latvia', 'Europe', 'Latvian'),
 ('Lesotho', 'Africa', 'Sotho'),
 ('Luxembourg', 'Europe', 'Luxembourgish'),
 ('Malawi', 'Africa', 'Chichewa'),
 ('Maldives', 'Asia', 'Dhivehi'),
 ('Mali', 'Africa', 'Bambara'),
 ('Malta', 'Europe', 'Maltese'),
 ('Marshall Islands', 'Oceania', 'Marshallese'),
 ('Mauritania', 'Africa', 'Hassaniya'),
 ('Mayotte', 'Africa', 'Mahoré'),
 ('Micronesia, Federated States of', 'Oceania', 'Trukese'),
 ('Mozambique', 'Africa', 'Makua'),
 ('Myanmar', 'Asia', 'Burmese'),
 ('Namibia', 'Africa', 'Ovambo'),
 ('Nauru', 'Oceania', 'Nauru'),
 ('Niue', 'Oceania', 'Niue'),
 ('Palau', 'Oceania', 'Palau'),
 ('Pitcairn', 'Oceania', 'Pitcairnese'),
 ('French Guiana', 'South America', 'Creole French'),
 ('French Polynesia', 'Oceania', 'Tahitian'),
 ('Zambia', 'Africa', 'Bemba'),
 ('Samoa', 'Oceania', 'Samoan-English'),
 ('Seychelles', 'Africa', 'Seselwa'),
 ('Sierra Leone', 'Africa', 'Mende'),
 ('Sri Lanka', 'Asia', 'Singali'),
 ('Suriname', 'South America', 'Sranantonga'),
 ('Taiwan', 'Asia', 'Min'),
 ('Tanzania', 'Africa', 'Nyamwesi'),
 ('Tokelau', 'Oceania', 'Tokelau'),
 ('Uganda', 'Africa', 'Ganda'),
 ('Wallis and Futuna', 'Oceania', 'Wallis'),
 ('Vanuatu', 'Oceania', 'Bislama'),
 ('Afghanistan', 'Asia', 'Dari'),
 ('Netherlands', 'Europe', 'Fries'),
 ('Angola', 'Africa', 'Mbundu'),
 ('Australia', 'Oceania', 'Italian'),
 ('Bangladesh', 'Asia', 'Chakma'),
 ('United Kingdom', 'Europe', 'Kymri'),
 ('Brunei', 'Asia', 'Malay-English'),
 ('Chile', 'South America', 'Araucan'),
 ('Egypt', 'Africa', 'Sinaberberi'),
 ('El Salvador', 'North America', 'Nahua'),
 ('Eritrea', 'Africa', 'Tigre'),
 ('South Africa', 'Africa', 'Xhosa'),
 ('Ethiopia', 'Africa', 'Amhara'),
 ('Fiji Islands', 'Oceania', 'Hindi'),
 ('Philippines', 'Asia', 'Cebuano'),
 ('Gabon', 'Africa', 'Punu-sira-nzebi'),
 ('Greenland', 'North America', 'Danish'),
 ('Guatemala', 'North America', 'Quiché'),
 ('Ireland', 'Europe', 'Irish'),
 ('Italy', 'Europe', 'Sardinian'),
 ('Yemen', 'Asia', 'Soqutri'),
 ('Jordan', 'Asia', 'Circassian'),
 ('Cameroon', 'Africa', 'Bamileke-bamum'),
 ('Kenya', 'Africa', 'Luhya'),
 ('Central African Republic', 'Africa', 'Banda'),
 ('China', 'Asia', 'Zhuang'),
 ('Colombia', 'South America', 'Chibcha'),
 ('Comoros', 'Africa', 'Comorian-French'),
 ('Laos', 'Asia', 'Mon-khmer'),
 ('Liberia', 'Africa', 'Bassa'),
 ('Mauritius', 'Africa', 'Bhojpuri'),
 ('Mexico', 'North America', 'Náhuatl'),
 ('Micronesia, Federated States of', 'Oceania', 'Pohnpei'),
 ('Monaco', 'Europe', 'Monegasque'),
 ('Myanmar', 'Asia', 'Shan'),
 ('Namibia', 'Africa', 'Nama'),
 ('Nepal', 'Asia', 'Maithili'),
 ('Niger', 'Africa', 'Songhai-zerma'),
 ('Côte d\x92Ivoire', 'Africa', 'Gur'),
 ('Equatorial Guinea', 'Africa', 'Bubi'),
 ('Réunion', 'Africa', 'Chinese'),
 ('Zambia', 'Africa', 'Tongan'),
 ('Sierra Leone', 'Africa', 'Temne'),
 ('Sudan', 'Africa', 'Dinka'),
 ('Suriname', 'South America', 'Hindi'),
 ('Togo', 'Africa', 'Kabyé'),
 ('Czech Republic', 'Europe', 'Moravian'),
 ('Tunisia', 'Africa', 'Arabic-French'),
 ('Uganda', 'Africa', 'Nkole'),
 ('Wallis and Futuna', 'Oceania', 'Futuna'),
 ('Venezuela', 'South America', 'Go

<br><br><br>

---

**Question:** Can you think of two advantages of a temporary table over CTEs?

---

<br><br><br>

Besides simplifying long and complex queries and increase readability, we can do things with CTEs that are impossible otherwise. One example is with grouping and aggregation.

Remember that when aggregate functions are used, only columns can be in the `SELECT` clause that are either aggregated (used inside an aggregate function) or appear also in the `GROUP BY` clause. Now, consider the following example:

<br><br><br>

---

**Example:** List city names, their country and region where they are located. Also in each row (i.e. for each city)

- list the number of official languages and
- number of cities with population over 1,000,000

in their respective countries.

---

We can obviously join the three tables `country`, `city`, and `countrylanguage` and do grouping, but we won't be able to list city names for the reason just discussed. But we can actually do this using a CTE:

In [14]:
%%sql

WITH
    ccl (code, country, region, "num_cities_over_1_mil", "num_off_lang")
AS (
    SELECT
        co.code, co.name, co.region,
        COUNT(DISTINCT ci.name),
        COUNT(DISTINCT cl.language)
    FROM
        country co
    JOIN
        city ci
    ON
        co.code = ci.countrycode
    JOIN
        countrylanguage cl
    ON
        co.code = cl.countrycode
    WHERE
        ci.population > 1000000
    GROUP BY
        co.code, co.name, co.region
)
SELECT
    ci.name, ccl.*
FROM
    ccl
JOIN
    city ci
ON
    ccl.code = ci.countrycode
;

   postgresql://postgres:***@localhost/imdb
 * postgresql://postgres:***@localhost:5432/world
3671 rows affected.


[('Kabul', 'AFG', 'Afghanistan', 'Southern and Central Asia', 1, 5),
 ('Qandahar', 'AFG', 'Afghanistan', 'Southern and Central Asia', 1, 5),
 ('Herat', 'AFG', 'Afghanistan', 'Southern and Central Asia', 1, 5),
 ('Mazar-e-Sharif', 'AFG', 'Afghanistan', 'Southern and Central Asia', 1, 5),
 ('Alger', 'DZA', 'Algeria', 'Northern Africa', 1, 2),
 ('Oran', 'DZA', 'Algeria', 'Northern Africa', 1, 2),
 ('Constantine', 'DZA', 'Algeria', 'Northern Africa', 1, 2),
 ('Annaba', 'DZA', 'Algeria', 'Northern Africa', 1, 2),
 ('Batna', 'DZA', 'Algeria', 'Northern Africa', 1, 2),
 ('Sétif', 'DZA', 'Algeria', 'Northern Africa', 1, 2),
 ('Sidi Bel Abbès', 'DZA', 'Algeria', 'Northern Africa', 1, 2),
 ('Skikda', 'DZA', 'Algeria', 'Northern Africa', 1, 2),
 ('Biskra', 'DZA', 'Algeria', 'Northern Africa', 1, 2),
 ('Blida (el-Boulaida)', 'DZA', 'Algeria', 'Northern Africa', 1, 2),
 ('Béjaïa', 'DZA', 'Algeria', 'Northern Africa', 1, 2),
 ('Mostaganem', 'DZA', 'Algeria', 'Northern Africa', 1, 2),
 ('Tébessa', 'DZA', 'Algeria', 'Northern Africa', 1, 2),
 ('Tlemcen (Tilimsen)', 'DZA', 'Algeria', 'Northern Africa', 1, 2),
 ('Béchar', 'DZA', 'Algeria', 'Northern Africa', 1, 2),
 ('Tiaret', 'DZA', 'Algeria', 'Northern Africa', 1, 2),
 ('Ech-Chleff (el-Asnam)', 'DZA', 'Algeria', 'Northern Africa', 1, 2),
 ('Ghardaïa', 'DZA', 'Algeria', 'Northern Africa', 1, 2),
 ('Luanda', 'AGO', 'Angola', 'Central Africa', 1, 9),
 ('Huambo', 'AGO', 'Angola', 'Central Africa', 1, 9),
 ('Lobito', 'AGO', 'Angola', 'Central Africa', 1, 9),
 ('Benguela', 'AGO', 'Angola', 'Central Africa', 1, 9),
 ('Namibe', 'AGO', 'Angola', 'Central Africa', 1, 9),
 ('Buenos Aires', 'ARG', 'Argentina', 'South America', 3, 3),
 ('La Matanza', 'ARG', 'Argentina', 'South America', 3, 3),
 ('Córdoba', 'ARG', 'Argentina', 'South America', 3, 3),
 ('Rosario', 'ARG', 'Argentina', 'South America', 3, 3),
 ('Lomas de Zamora', 'ARG', 'Argentina', 'South America', 3, 3),
 ('Quilmes', 'ARG', 'Argentina', 'South America', 3, 3),
 ('Almirante Brown', 'ARG', 'Argentina', 'South America', 3, 3),
 ('La Plata', 'ARG', 'Argentina', 'South America', 3, 3),
 ('Mar del Plata', 'ARG', 'Argentina', 'South America', 3, 3),
 ('San Miguel de Tucumán', 'ARG', 'Argentina', 'South America', 3, 3),
 ('Lanús', 'ARG', 'Argentina', 'South America', 3, 3),
 ('Merlo', 'ARG', 'Argentina', 'South America', 3, 3),
 ('General San Martín', 'ARG', 'Argentina', 'South America', 3, 3),
 ('Salta', 'ARG', 'Argentina', 'South America', 3, 3),
 ('Moreno', 'ARG', 'Argentina', 'South America', 3, 3),
 ('Santa Fé', 'ARG', 'Argentina', 'South America', 3, 3),
 ('Avellaneda', 'ARG', 'Argentina', 'South America', 3, 3),
 ('Tres de Febrero', 'ARG', 'Argentina', 'South America', 3, 3),
 ('Morón', 'ARG', 'Argentina', 'South America', 3, 3),
 ('Florencio Varela', 'ARG', 'Argentina', 'South America', 3, 3),
 ('San Isidro', 'ARG', 'Argentina', 'South America', 3, 3),
 ('Tigre', 'ARG', 'Argentina', 'South America', 3, 3),
 ('Malvinas Argentinas', 'ARG', 'Argentina', 'South America', 3, 3),
 ('Vicente López', 'ARG', 'Argentina', 'South America', 3, 3),
 ('Berazategui', 'ARG', 'Argentina', 'South America', 3, 3),
 ('Corrientes', 'ARG', 'Argentina', 'South America', 3, 3),
 ('San Miguel', 'ARG', 'Argentina', 'South America', 3, 3),
 ('Bahía Blanca', 'ARG', 'Argentina', 'South America', 3, 3),
 ('Esteban Echeverría', 'ARG', 'Argentina', 'South America', 3, 3),
 ('Resistencia', 'ARG', 'Argentina', 'South America', 3, 3),
 ('José C. Paz', 'ARG', 'Argentina', 'South America', 3, 3),
 ('Paraná', 'ARG', 'Argentina', 'South America', 3, 3),
 ('Godoy Cruz', 'ARG', 'Argentina', 'South America', 3, 3),
 ('Posadas', 'ARG', 'Argentina', 'South America', 3, 3),
 ('Guaymallén', 'ARG', 'Argentina', 'South America', 3, 3),
 ('Santiago del Estero', 'ARG', 'Argentina', 'South America', 3, 3),
 ('San Salvador de Jujuy', 'ARG', 'Argentina', 'South America', 3, 3),
 ('Hurlingham', 'ARG', 'Argentina', 'South America', 3, 3),
 ('Neuquén', 'ARG', 'Argentina', 'South America', 3, 3),

<br><br><br>

### (OPTIONAL) Data-modifying CTEs

CTEs can also be used with `INSERT`, `DELETE`, and `UPDATE` to catch the rows modified by these statements via the `RETURNING` statement and use them for another purpose.

Let's connect to the `mds` database we created earlier and reload it with the original data:

In [15]:
%sql postgresql://{username}:{password}@{host}:{port}/mds

'Connected: postgres@mds'

In [16]:
%%sql

DROP TABLE IF EXISTS
    instructor,
    instructor_course,
    course_cohort
;

CREATE TABLE instructor (
    id INTEGER PRIMARY KEY,
    name TEXT,
    email TEXT,
    phone VARCHAR(12),
    department VARCHAR(50)
    )
;

INSERT INTO
    instructor (id, name, email, phone, department)
VALUES
    (1, 'Mike', 'mike@mds.ubc.ca', '605-332-2343', 'Computer Science'),
    (2, 'Tiffany', 'tiff@mds.ubc.ca', '445-794-2233', 'Neuroscience'),
    (3, 'Arman', 'arman@mds.ubc.ca', '935-738-5796', 'Physics'),
    (4, 'Varada', 'varada@mds.ubc.ca', '243-924-4446', 'Computer Science'),
    (5, 'Quan', 'quan@mds.ubc.ca', '644-818-0254', 'Economics'),
    (6, 'Joel', 'joel@mds.ubc.ca', '773-432-7669', 'Biomedical Engineering'),
    (7, 'Florencia', 'flor@mds.ubc.ca', '773-926-2837', 'Biology'),
    (8, 'Alexi', 'alexiu@mds.ubc.ca', '421-888-4550', 'Statistics'),
    (15, 'Vincenzo', 'vincenzo@mds.ubc.ca', '776-543-1212', 'Statistics'),
    (19, 'Gittu', 'gittu@mds.ubc.ca', '776-334-1132', 'Biomedical Engineering'),
    (16, 'Jessica', 'jessica@mds.ubc.ca', '211-990-1762', 'Computer Science')
;

    
CREATE TABLE instructor_course (
    id SERIAL PRIMARY KEY,
    instructor_id INTEGER,
    course TEXT,
    enrollment INTEGER,
    begins DATE
    )
;

INSERT INTO
    instructor_course (instructor_id, course, enrollment, begins)
VALUES
    (8, 'Statistical Inference and Computation I', 125, '2021-10-01'),
    (8, 'Regression II', 102, '2022-02-05'),
    (1, 'Descriptive Statistics and Probability', 79, '2021-09-10'),
    (1, 'Algorithms and Data Structures', 25, '2021-10-01'),
    (3, 'Algorithms and Data Structures', 25, '2021-10-01'),
    (3, 'Python Programming', 133, '2021-09-07'),
    (3, 'Databases & Data Retrieval', 118, '2021-11-16'),
    (6, 'Visualization I', 155, '2021-10-01'),
    (6, 'Privacy, Ethics & Security', 148, '2022-03-01'),
    (2, 'Programming for Data Manipulation', 160, '2021-09-08'),
    (7, 'Data Science Workflows', 98, '2021-09-15'),
    (2, 'Data Science Workflows', 98, '2021-09-15'),
    (12, 'Web & Cloud Computing', 78, '2022-02-10'),
    (10, 'Introduction to Optimization', NULL, '2022-09-01'),
    (9, 'Parallel Computing', NULL, '2023-01-10'),
    (13, 'Natural Language Processing', NULL, '2023-09-10')
;

CREATE TABLE course_cohort (
    id INTEGER,
    cohort VARCHAR(7)
    )
;

INSERT INTO
    course_cohort (id, cohort)
VALUES
    (13, 'MDS-CL'),
    (8, 'MDS-CL'),
    (1, 'MDS-CL'),
    (3, 'MDS-CL'),
    (1, 'MDS-V'),
    (9, 'MDS-V'),
    (9, 'MDS-V'),
    (3, 'MDS-V')
;

   postgresql://postgres:***@localhost/imdb
 * postgresql://postgres:***@localhost:5432/mds
   postgresql://postgres:***@localhost:5432/world
Done.
Done.
11 rows affected.
Done.
16 rows affected.
Done.
8 rows affected.


[]

In [17]:
%sql SELECT * FROM instructor;

   postgresql://postgres:***@localhost/imdb
 * postgresql://postgres:***@localhost:5432/mds
   postgresql://postgres:***@localhost:5432/world
11 rows affected.


[(1, 'Mike', 'mike@mds.ubc.ca', '605-332-2343', 'Computer Science'),
 (2, 'Tiffany', 'tiff@mds.ubc.ca', '445-794-2233', 'Neuroscience'),
 (3, 'Arman', 'arman@mds.ubc.ca', '935-738-5796', 'Physics'),
 (4, 'Varada', 'varada@mds.ubc.ca', '243-924-4446', 'Computer Science'),
 (5, 'Quan', 'quan@mds.ubc.ca', '644-818-0254', 'Economics'),
 (6, 'Joel', 'joel@mds.ubc.ca', '773-432-7669', 'Biomedical Engineering'),
 (7, 'Florencia', 'flor@mds.ubc.ca', '773-926-2837', 'Biology'),
 (8, 'Alexi', 'alexiu@mds.ubc.ca', '421-888-4550', 'Statistics'),
 (15, 'Vincenzo', 'vincenzo@mds.ubc.ca', '776-543-1212', 'Statistics'),
 (19, 'Gittu', 'gittu@mds.ubc.ca', '776-334-1132', 'Biomedical Engineering'),
 (16, 'Jessica', 'jessica@mds.ubc.ca', '211-990-1762', 'Computer Science')]

In [18]:
%%sql

DROP TABLE IF EXISTS former_instructor;

CREATE TABLE
    former_instructor AS (
    SELECT
        *
    FROM
        instructor
    WHERE
        1 = 2
)
;

   postgresql://postgres:***@localhost/imdb
 * postgresql://postgres:***@localhost:5432/mds
   postgresql://postgres:***@localhost:5432/world
Done.
0 rows affected.


[]

In [19]:
%sql SELECT * FROM former_instructor;

   postgresql://postgres:***@localhost/imdb
 * postgresql://postgres:***@localhost:5432/mds
   postgresql://postgres:***@localhost:5432/world
0 rows affected.


[]

Here I use the `RETURNING` statement to retrieve the results of deleting two rows using a CTE, and insert the deleted rows into another table, namely, `former_instructor`:

In [20]:
%%sql

WITH deleted_rows AS (
    DELETE FROM
        instructor
    WHERE
        name IN ('Arman', 'Mike')
    RETURNING
        *
)
INSERT INTO
    former_instructor
SELECT
    *
FROM
    deleted_rows
;

   postgresql://postgres:***@localhost/imdb
 * postgresql://postgres:***@localhost:5432/mds
   postgresql://postgres:***@localhost:5432/world
2 rows affected.


[]

Let's check out the `former_instructor` table now:

In [21]:
%sql SELECT * FROM former_instructor;

   postgresql://postgres:***@localhost/imdb
 * postgresql://postgres:***@localhost:5432/mds
   postgresql://postgres:***@localhost:5432/world
2 rows affected.


[(1, 'Mike', 'mike@mds.ubc.ca', '605-332-2343', 'Computer Science'),
 (3, 'Arman', 'arman@mds.ubc.ca', '935-738-5796', 'Physics')]

The two deleted rows have been successfully moved to the `former_instructor` table!

<br><br><br>

## Window functions

We've been using aggregation and grouping quite a lot so far. But as you may recall, there are situations that aggregation and grouping alone cannot address, such as having aggregated and non-aggregated columns together.

We know that aggregation functions perform calculations on groups of rows, and return a single value for all rows that take part in that calculation. They also collapse the participant rows into one single row with the the common column(s) values between all the collapsed rows and the aggregated value. For example, in the `instructor` table of the `mds` database, we can group rows by `department`, and apply `COUNT()` to find the number of instructors in each department, but we no longer see individual instructor rows. Instead, we will see a `COUNT()` number for each `department`.

Window functions act in a similar fashion: they perform operations on groups of rows that are related in some way to the current row (like `GROUP BY` that bases the relatedness on common values in grouped columns), but these functions **do not collapse rows**! This is the most important difference between aggregation and window functions

Let's take a look at an example first. The following query finds the maximum population of countries in each continent:

In [22]:
%sql postgresql://{username}:{password}@{host}:{port}/world

'Connected: postgres@world'

In [23]:
%%sql

SELECT
    continent,
    MAX(population)
FROM
    country
GROUP BY
    continent
;

   postgresql://postgres:***@localhost/imdb
   postgresql://postgres:***@localhost:5432/mds
 * postgresql://postgres:***@localhost:5432/world
7 rows affected.


[('Asia', 1277558000),
 ('South America', 170115000),
 ('North America', 278357000),
 ('Oceania', 18886000),
 ('Antarctica', 0),
 ('Africa', 111506000),
 ('Europe', 146934000)]

<br><br><br>

However, what if we want each country to be listed along with its own population, as well as the maximum population in their respective continent? This is not possible with aggregation functions alone.

In order to list country names, their population, and the maximum population in their continent we can use `MAX()` as a window function:

In [24]:
%%sql

SELECT
    name,
    population,
    continent,
    MAX(population)
        OVER (PARTITION BY continent)
FROM
    country
;

   postgresql://postgres:***@localhost/imdb
   postgresql://postgres:***@localhost:5432/mds
 * postgresql://postgres:***@localhost:5432/world
239 rows affected.


[('Algeria', 31471000, 'Africa', 111506000),
 ('Western Sahara', 293000, 'Africa', 111506000),
 ('Madagascar', 15942000, 'Africa', 111506000),
 ('Uganda', 21778000, 'Africa', 111506000),
 ('Malawi', 10925000, 'Africa', 111506000),
 ('Mali', 11234000, 'Africa', 111506000),
 ('Morocco', 28351000, 'Africa', 111506000),
 ('Côte d\x92Ivoire', 14786000, 'Africa', 111506000),
 ('Mauritania', 2670000, 'Africa', 111506000),
 ('Mauritius', 1158000, 'Africa', 111506000),
 ('Mayotte', 149000, 'Africa', 111506000),
 ('Nigeria', 111506000, 'Africa', 111506000),
 ('Niger', 10730000, 'Africa', 111506000),
 ('Mozambique', 19680000, 'Africa', 111506000),
 ('Namibia', 1726000, 'Africa', 111506000),
 ('Botswana', 1622000, 'Africa', 111506000),
 ('Tunisia', 9586000, 'Africa', 111506000),
 ('Angola', 12878000, 'Africa', 111506000),
 ('Burkina Faso', 11937000, 'Africa', 111506000),
 ('Burundi', 6695000, 'Africa', 111506000),
 ('Chad', 7651000, 'Africa', 111506000),
 ('Togo', 4629000, 'Africa', 111506000),
 ('Tanzania', 33517000, 'Africa', 111506000),
 ('Djibouti', 638000, 'Africa', 111506000),
 ('Egypt', 68470000, 'Africa', 111506000),
 ('Swaziland', 1008000, 'Africa', 111506000),
 ('Eritrea', 3850000, 'Africa', 111506000),
 ('South Africa', 40377000, 'Africa', 111506000),
 ('Ethiopia', 62565000, 'Africa', 111506000),
 ('Sudan', 29490000, 'Africa', 111506000),
 ('British Indian Ocean Territory', 0, 'Africa', 111506000),
 ('Gabon', 1226000, 'Africa', 111506000),
 ('Gambia', 1305000, 'Africa', 111506000),
 ('Ghana', 20212000, 'Africa', 111506000),
 ('Somalia', 10097000, 'Africa', 111506000),
 ('Sierra Leone', 4854000, 'Africa', 111506000),
 ('Seychelles', 77000, 'Africa', 111506000),
 ('Senegal', 9481000, 'Africa', 111506000),
 ('Guinea', 7430000, 'Africa', 111506000),
 ('Guinea-Bissau', 1213000, 'Africa', 111506000),
 ('Sao Tome and Principe', 147000, 'Africa', 111506000),
 ('Zambia', 9169000, 'Africa', 111506000),
 ('Saint Helena', 6000, 'Africa', 111506000),
 ('Zimbabwe', 11669000, 'Africa', 111506000),
 ('Cameroon', 15085000, 'Africa', 111506000),
 ('Rwanda', 7733000, 'Africa', 111506000),
 ('Cape Verde', 428000, 'Africa', 111506000),
 ('Kenya', 30080000, 'Africa', 111506000),
 ('Central African Republic', 3615000, 'Africa', 111506000),
 ('Réunion', 699000, 'Africa', 111506000),
 ('Comoros', 578000, 'Africa', 111506000),
 ('Congo', 2943000, 'Africa', 111506000),
 ('Congo, The Democratic Republic of the', 51654000, 'Africa', 111506000),
 ('Equatorial Guinea', 453000, 'Africa', 111506000),
 ('Lesotho', 2153000, 'Africa', 111506000),
 ('Liberia', 3154000, 'Africa', 111506000),
 ('Libyan Arab Jamahiriya', 5605000, 'Africa', 111506000),
 ('Benin', 6097000, 'Africa', 111506000),
 ('Heard Island and McDonald Islands', 0, 'Antarctica', 0),
 ('Bouvet Island', 0, 'Antarctica', 0),
 ('French Southern territories', 0, 'Antarctica', 0),
 ('Antarctica', 0, 'Antarctica', 0),
 ('South Georgia and the South Sandwich Islands', 0, 'Antarctica', 0),
 ('Pakistan', 156483000, 'Asia', 1277558000),
 ('China', 1277558000, 'Asia', 1277558000),
 ('Maldives', 286000, 'Asia', 1277558000),
 ('Malaysia', 22244000, 'Asia', 1277558000),
 ('Kyrgyzstan', 4699000, 'Asia', 1277558000),
 ('Bhutan', 2124000, 'Asia', 1277558000),
 ('Kuwait', 1972000, 'Asia', 1277558000),
 ('Oman', 2542000, 'Asia', 1277558000),
 ('Qatar', 599000, 'Asia', 1277558000),
 ('Saudi Arabia', 21607000, 'Asia', 1277558000),
 ('Cyprus', 754700, 'Asia', 1277558000),
 ('Afghanistan', 22720000, 'Asia', 1277558000),
 ('Hong Kong', 6782000, 'Asia', 1277558000),
 ('Turkmenistan', 4459000, 'Asia', 1277558000),
 ('Mongolia', 2662000, 'Asia', 1277558000),
 ('Nepal', 23930000, 'Asia', 1277558000),
 ('Indonesia', 212107000, 'Asia', 1277558000),
 ('Myanmar', 45611000, 'Asia', 1277558000),
 ('India', 1013662000, 'Asia', 1277558000),
 ('Iraq', 23115000, 'Asia', 1277558000),
 ('Turkey', 66591000, 'Asia', 1277558000),
 ('Iran', 67702000, 'Asia', 1277558000),
 ('Brunei', 328000, 'Asia', 1277558000),
 ('Armenia', 3520000, 'Asia', 1

<br><br><br>

### Syntax

A lot of new keywords appeared all at once here, so let me explain them all in detail. The syntax of a window function is:
‍‍‍
```sql
SELECT
    column,
    .
    .
    .
    window_function_name(expression) OVER (
        PARTITION BY column
        ORDER BY column
        frame_clause
    .
    .
    .
FROM
    table
;
```

- `window_function_name()` specifies which window function we want to use. I'll list the different types of window functions in a bit.

- `OVER` is what tells the SQL engine that we are going to use a window function

- `PARTITION BY` determines how or based on which column we want the window function to subdivide the rows into groups

- `ORDER BY` specifies how we want the rows in each subgroup to be ordered

- `frame_clause` gives us further control over which rows in each subgroup can participate in the window function calculations

<br><br><br>

> **Note:** Window functions in SQL are processed **after** `GROUP BY` and `HAVING`, and **before** `SELECT`. This is a crucial point to understand when using window functions in SQL.

> **Note:** Window functions are only allowed in the `SELECT` and `ORDER BY` clauses.

<br><br><br>

### Types of window functions

<br><br><br>

#### Aggregate functions

These are the regular aggregate functions that we've learned about before, i.e. `AVG()`, `COUNT()`, `MAX()`, `MIN()`, and `SUM()`, which can also be used as window functions.

<br><br><br>

#### Ranking functions

- `CUME_DIST()`: returns the cumulative distribution, i.e. the percentage of values less than or equal to the current value.
- `NTILE()`: given a specified number of buckets, it tells us in which bucket each row goes among other rows in the partition.
- `PERCENT_RANK()`: similar to `CUME_DIST()`, but considers only the percentage of values less than (and not equal to) the current value.

- `DENSE_RANK()`: returns the rank of a row within a partition without jumps after duplicate ranks (e.g. 1, 2, 2, 3, ...)
- `RANK()`: returns the rank of row within a partition but with jumps after duplicates ranks (e.g. 1, 2, 2, 4, ...)
- `ROW_NUMBER()`: returns simply the number of a row in a partition, regardless of duplicate values

<br><br><br>

#### Value functions

- `FIRST_VALUE()`: returns the first value of the ordered partition

- `LAST_VALUE()`: returns the last value of the ordered partition
- `LAG()`: returns value from the preceding row with a specified offset
- `LEAD()`: returns value from the following row with a specified offset

<br><br><br>

**Differences with subqueries**

- Subqueries can operate on multiple tables, window functions are limited to the current table

- Subqueries are far more general (e.g. `WHERE` conditions including `EXITS`, etc.)
- Window functions are more readable
- Window functions offer functionalities such as `RANK()` or `LAG()` and `LEAD()` that are either very hard, if not impossible, to implement using subqueries
- Window functions are generally faster than subqueries when used for a same purpose

<br><br><br>

---

**Example:** Using window functions, count the number of countries in which each language is officially spoken. Remove duplicates and sort the results in descending order by the count value.

---

In [25]:
%%sql

SELECT
    DISTINCT language,
    COUNT(*) OVER (PARTITION BY language) AS count
FROM
    countrylanguage
WHERE
    isofficial = TRUE
ORDER BY
    count DESC

   postgresql://postgres:***@localhost/imdb
   postgresql://postgres:***@localhost:5432/mds
 * postgresql://postgres:***@localhost:5432/world
102 rows affected.


[('English', 44),
 ('Arabic', 22),
 ('Spanish', 20),
 ('French', 18),
 ('German', 6),
 ('Portuguese', 6),
 ('Dutch', 4),
 ('Italian', 4),
 ('Malay', 4),
 ('Danish', 3),
 ('Russian', 3),
 ('Serbo-Croatian', 3),
 ('Aimará', 2),
 ('Chinese', 2),
 ('Greek', 2),
 ('Ket\x9aua', 2),
 ('Korean', 2),
 ('Norwegian', 2),
 ('Romanian', 2),
 ('Samoan', 2),
 ('Swedish', 2),
 ('Tamil', 2),
 ('Turkish', 2),
 ('Afrikaans', 1),
 ('Albaniana', 1),
 ('Armenian', 1),
 ('Azerbaijani', 1),
 ('Belorussian', 1),
 ('Bengali', 1),
 ('Bislama', 1),
 ('Bulgariana', 1),
 ('Burmese', 1),
 ('Catalan', 1),
 ('Chamorro', 1),
 ('Chichewa', 1),
 ('Comorian', 1),
 ('Czech', 1),
 ('Dari', 1),
 ('Dhivehi', 1),
 ('Dzongkha', 1),
 ('Estonian', 1),
 ('Ewe', 1),
 ('Faroese', 1),
 ('Fijian', 1),
 ('Finnish', 1),
 ('Georgiana', 1),
 ('Greenlandic', 1),
 ('Guaraní', 1),
 ('Hebrew', 1),
 ('Hindi', 1),
 ('Hungarian', 1),
 ('Icelandic', 1),
 ('Irish', 1),
 ('Japanese', 1),
 ('Kabyé', 1),
 ('Kazakh', 1),
 ('Khmer', 1),
 ('Kirgiz', 1),
 ('Kiribati', 1),
 ('Kirundi', 1),
 ('Lao', 1),
 ('Latvian', 1),
 ('Lithuanian', 1),
 ('Luxembourgish', 1),
 ('Macedonian', 1),
 ('Malagasy', 1),
 ('Maltese', 1),
 ('Mandarin Chinese', 1),
 ('Maori', 1),
 ('Marshallese', 1),
 ('Mongolian', 1),
 ('Nauru', 1),
 ('Nepali', 1),
 ('Palau', 1),
 ('Papiamento', 1),
 ('Pashto', 1),
 ('Persian', 1),
 ('Pilipino', 1),
 ('Polish', 1),
 ('Romani', 1),
 ('Romansh', 1),
 ('Rwanda', 1),
 ('Singali', 1),
 ('Slovak', 1),
 ('Slovene', 1),
 ('Somali', 1),
 ('Sotho', 1),
 ('Swahili', 1),
 ('Swazi', 1),
 ('Tadzhik', 1),
 ('Thai', 1),
 ('Tigrinja', 1),
 ('Tongan', 1),
 ('Turkmenian', 1),
 ('Tuvalu', 1),
 ('Ukrainian', 1),
 ('Urdu', 1),
 ('Uzbek', 1),
 ('Vietnamese', 1),
 ('Wolof', 1),
 ('Xhosa', 1),
 ('Zulu', 1)]

<br><br><br>

**Question:** In the above query, is it possible to use a `WHERE` clause to limit the results to languages which are spoken in more than one country?

<br><br><br>

In the following example, we rank countries by life expectancy in each continent, but we also create a column to retrieve the name of the country whose ranking follows the one in the current row:

In [26]:
%%sql

SELECT
    name,
    continent,
    DENSE_RANK()
        OVER (PARTITION BY continent ORDER BY lifeexpectancy DESC),
    LEAD(name, 1)
        OVER (PARTITION BY continent ORDER BY lifeexpectancy DESC)
FROM
    country
WHERE
    population > 10000000

   postgresql://postgres:***@localhost/imdb
   postgresql://postgres:***@localhost:5432/mds
 * postgresql://postgres:***@localhost:5432/world
78 rows affected.


[('Algeria', 'Africa', 1, 'Morocco'),
 ('Morocco', 'Africa', 2, 'Egypt'),
 ('Egypt', 'Africa', 3, 'Ghana'),
 ('Ghana', 'Africa', 4, 'Sudan'),
 ('Sudan', 'Africa', 5, 'Madagascar'),
 ('Madagascar', 'Africa', 6, 'Cameroon'),
 ('Cameroon', 'Africa', 7, 'Tanzania'),
 ('Tanzania', 'Africa', 8, 'Nigeria'),
 ('Nigeria', 'Africa', 9, 'South Africa'),
 ('South Africa', 'Africa', 10, 'Congo, The Democratic Republic of the'),
 ('Congo, The Democratic Republic of the', 'Africa', 11, 'Kenya'),
 ('Kenya', 'Africa', 12, 'Burkina Faso'),
 ('Burkina Faso', 'Africa', 13, 'Mali'),
 ('Mali', 'Africa', 13, 'Somalia'),
 ('Somalia', 'Africa', 14, 'Ethiopia'),
 ('Ethiopia', 'Africa', 15, 'Côte d\x92Ivoire'),
 ('Côte d\x92Ivoire', 'Africa', 15, 'Uganda'),
 ('Uganda', 'Africa', 16, 'Niger'),
 ('Niger', 'Africa', 17, 'Angola'),
 ('Angola', 'Africa', 18, 'Zimbabwe'),
 ('Zimbabwe', 'Africa', 19, 'Malawi'),
 ('Malawi', 'Africa', 20, 'Mozambique'),
 ('Mozambique', 'Africa', 21, None),
 ('Japan', 'Asia', 1, 'Taiwan'),
 ('Taiwan', 'Asia', 2, 'South Korea'),
 ('South Korea', 'Asia', 3, 'Sri Lanka'),
 ('Sri Lanka', 'Asia', 4, 'China'),
 ('China', 'Asia', 5, 'Turkey'),
 ('Turkey', 'Asia', 6, 'Malaysia'),
 ('Malaysia', 'Asia', 7, 'North Korea'),
 ('North Korea', 'Asia', 8, 'Iran'),
 ('Iran', 'Asia', 9, 'Vietnam'),
 ('Vietnam', 'Asia', 10, 'Thailand'),
 ('Thailand', 'Asia', 11, 'Syria'),
 ('Syria', 'Asia', 12, 'Indonesia'),
 ('Indonesia', 'Asia', 13, 'Saudi Arabia'),
 ('Saudi Arabia', 'Asia', 14, 'Philippines'),
 ('Philippines', 'Asia', 15, 'Iraq'),
 ('Iraq', 'Asia', 16, 'Uzbekistan'),
 ('Uzbekistan', 'Asia', 17, 'Kazakstan'),
 ('Kazakstan', 'Asia', 18, 'India'),
 ('India', 'Asia', 19, 'Pakistan'),
 ('Pakistan', 'Asia', 20, 'Bangladesh'),
 ('Bangladesh', 'Asia', 21, 'Yemen'),
 ('Yemen', 'Asia', 22, 'Nepal'),
 ('Nepal', 'Asia', 23, 'Cambodia'),
 ('Cambodia', 'Asia', 24, 'Myanmar'),
 ('Myanmar', 'Asia', 25, 'Afghanistan'),
 ('Afghanistan', 'Asia', 26, None),
 ('Italy', 'Europe', 1, 'France'),
 ('France', 'Europe', 2, 'Spain'),
 ('Spain', 'Europe', 2, 'Greece'),
 ('Greece', 'Europe', 3, 'Netherlands'),
 ('Netherlands', 'Europe', 4, 'Belgium'),
 ('Belgium', 'Europe', 5, 'United Kingdom'),
 ('United Kingdom', 'Europe', 6, 'Germany'),
 ('Germany', 'Europe', 7, 'Czech Republic'),
 ('Czech Republic', 'Europe', 8, 'Poland'),
 ('Poland', 'Europe', 9, 'Yugoslavia'),
 ('Yugoslavia', 'Europe', 10, 'Hungary'),
 ('Hungary', 'Europe', 11, 'Romania'),
 ('Romania', 'Europe', 12, 'Belarus'),
 ('Belarus', 'Europe', 13, 'Russian Federation'),
 ('Russian Federation', 'Europe', 14, 'Ukraine'),
 ('Ukraine', 'Europe', 15, None),
 ('Canada', 'North America', 1, 'United States'),
 ('United States', 'North America', 2, 'Cuba'),
 ('Cuba', 'North America', 3, 'Mexico'),
 ('Mexico', 'North America', 4, 'Guatemala'),
 ('Guatemala', 'North America', 5, None),
 ('Australia', 'Oceania', 1, None),
 ('Chile', 'South America', 1, 'Argentina'),
 ('Argentina', 'South America', 2, 'Venezuela'),
 ('Venezuela', 'South America', 3, 'Ecuador'),
 ('Ecuador', 'South America', 4, 'Colombia'),
 ('Colombia', 'South America', 5, 'Peru'),
 ('Peru', 'South America', 6, 'Brazil'),
 ('Brazil', 'South America', 7, None)]

<br><br><br>

In [27]:
%%sql

WITH cte(name, continent, rank, lead_rank) AS
(
SELECT
    name,
    continent,
    DENSE_RANK()
        OVER (PARTITION BY continent ORDER BY lifeexpectancy DESC),
    LEAD(name, 1)
        OVER (PARTITION BY continent ORDER BY lifeexpectancy DESC)
FROM
    country
WHERE
    population > 10000000
)
SELECT
    *
FROM
    cte
WHERE
    rank = 1

   postgresql://postgres:***@localhost/imdb
   postgresql://postgres:***@localhost:5432/mds
 * postgresql://postgres:***@localhost:5432/world
6 rows affected.


[('Algeria', 'Africa', 1, 'Morocco'),
 ('Japan', 'Asia', 1, 'Taiwan'),
 ('Italy', 'Europe', 1, 'France'),
 ('Canada', 'North America', 1, 'United States'),
 ('Australia', 'Oceania', 1, None),
 ('Chile', 'South America', 1, 'Argentina')]

<br><br><br>

---

**Example:** Write a query that in each row returns each movie's name, production year (`start_year`), and its number of votes, for all movies produced in or after 2018. Your query should also return a column that ranks each movie in descending order, according to its number of votes in the year it was produced.

- Sort your results according the rank value in ascending order

- Retrieve only 20 rows

---

In [28]:
%sql postgresql://{username}:{password}@{host}:{port}/imdb

'Connected: postgres@imdb'

In [29]:
%%sql

SELECT
    title,
    start_year,
    nvotes,
    RANK()
        OVER (PARTITION BY start_year ORDER BY nvotes DESC)
            AS rank
FROM
    movies
WHERE
    start_year >= 2018
ORDER BY
    rank
LIMIT
    20
;

   postgresql://postgres:***@localhost/imdb
 * postgresql://postgres:***@localhost:5432/imdb
   postgresql://postgres:***@localhost:5432/mds
   postgresql://postgres:***@localhost:5432/world
20 rows affected.


[('Avengers: Infinity War', 2018, 711500, 1),
 ('Avengers: Endgame', 2019, 567968, 1),
 ('Black Panther', 2018, 543002, 2),
 ('Captain Marvel', 2019, 362609, 2),
 ('Once Upon a Time... in Hollywood', 2019, 204266, 3),
 ('Deadpool 2', 2018, 416811, 3),
 ('Bohemian Rhapsody', 2018, 382622, 4),
 ('Spider-Man: Far from Home', 2019, 198519, 4),
 ('A Quiet Place', 2018, 325428, 5),
 ('Shazam!', 2019, 182200, 5),
 ('John Wick: Chapter 3 - Parabellum', 2019, 172900, 6),
 ('Ready Player One', 2018, 316326, 6),
 ('Alita: Battle Angel', 2019, 162513, 7),
 ('Venom', 2018, 305111, 7),
 ('Glass', 2019, 159064, 8),
 ('Aquaman', 2018, 300064, 8),
 ('Us', 2019, 150951, 9),
 ('A Star Is Born', 2018, 276484, 9),
 ('Spider-Man: Into the Spider-Verse', 2018, 262505, 10),
 ('Aladdin', 2019, 139370, 10)]

<br><br><br>

---

**Example:** We want to find out how movies are ranked in their respective genres. Write a query that in each row returns movie name, genre, and production year (`start_year`) for movies with more than 1,000,000 votes. Your query should also have a column that shows the ranking of each movie according to its rating in descending order.

- Sort your results by genre and rank in ascending order
- Limit your results to 50 rows

---

In [30]:
%%sql

SELECT
    title,
    genre,
    start_year,
    RANK()
        OVER (PARTITION BY genre ORDER BY rating)
            AS rank
FROM
    movies m
JOIN
    movie_genres mg
ON
    m.id = mg.movie_id
WHERE
    nvotes > 1000000
ORDER BY
    genre, rank
LIMIT
    50
;

   postgresql://postgres:***@localhost/imdb
 * postgresql://postgres:***@localhost:5432/imdb
   postgresql://postgres:***@localhost:5432/mds
   postgresql://postgres:***@localhost:5432/world
50 rows affected.


[('Avatar', 'action', 2009, 1),
 ('The Avengers', 'action', 2012, 2),
 ('Batman Begins', 'action', 2005, 3),
 ('The Dark Knight Rises', 'action', 2012, 4),
 ('Gladiator', 'action', 2000, 5),
 ('Star Wars: Episode IV - A New Hope', 'action', 1977, 6),
 ('The Matrix', 'action', 1999, 7),
 ('Star Wars: Episode V - The Empire Strikes Back', 'action', 1980, 7),
 ('Inception', 'action', 2010, 9),
 ('The Dark Knight', 'action', 2008, 10),
 ('Avatar', 'adventure', 2009, 1),
 ('The Avengers', 'adventure', 2012, 2),
 ('Batman Begins', 'adventure', 2005, 3),
 ('Inglourious Basterds', 'adventure', 2009, 4),
 ('Gladiator', 'adventure', 2000, 5),
 ('Interstellar', 'adventure', 2014, 6),
 ('Star Wars: Episode IV - A New Hope', 'adventure', 1977, 6),
 ('Star Wars: Episode V - The Empire Strikes Back', 'adventure', 1980, 8),
 ('The Lord of the Rings: The Two Towers', 'adventure', 2002, 8),
 ('The Lord of the Rings: The Fellowship of the Ring', 'adventure', 2001, 10),
 ('Inception', 'adventure', 2010, 10),
 ('The Lord of the Rings: The Return of the King', 'adventure', 2003, 12),
 ('The Wolf of Wall Street', 'biography', 2013, 1),
 ("Schindler's List", 'biography', 1993, 2),
 ('The Wolf of Wall Street', 'crime', 2013, 1),
 ('The Departed', 'crime', 2006, 2),
 ('Se7en', 'crime', 1995, 3),
 ('The Green Mile', 'crime', 1999, 3),
 ('The Silence of the Lambs', 'crime', 1991, 3),
 ('Pulp Fiction', 'crime', 1994, 6),
 ('The Godfather: Part II', 'crime', 1974, 7),
 ('The Dark Knight', 'crime', 2008, 7),
 ('The Godfather', 'crime', 1972, 9),
 ('The Wolf of Wall Street', 'drama', 2013, 1),
 ('Inglourious Basterds', 'drama', 2009, 2),
 ('American Beauty', 'drama', 1999, 2),
 ('Django Unchained', 'drama', 2012, 4),
 ('Gladiator', 'drama', 2000, 5),
 ('The Departed', 'drama', 2006, 5),
 ('The Prestige', 'drama', 2006, 5),
 ('The Green Mile', 'drama', 1999, 8),
 ('Saving Private Ryan', 'drama', 1998, 8),
 ('Se7en', 'drama', 1995, 8),
 ('The Silence of the Lambs', 'drama', 1991, 8),
 ('Interstellar', 'drama', 2014, 8),
 ('The Lord of the Rings: The Two Towers', 'drama', 2002, 13),
 ('Forrest Gump', 'drama', 1994, 14),
 ('Fight Club', 'drama', 1999, 14),
 ('The Lord of the Rings: The Fellowship of the Ring', 'drama', 2001, 14),
 ('The Lord of the Rings: The Return of the King', 'drama', 2003, 17)]

<br><br><br>

It is also possible to define multiple windows over which multiple window functions may operate:

In [31]:
%sql postgresql://{username}:{password}@{host}:{port}/world

'Connected: postgres@world'

In [32]:
%%sql

SELECT
    name,
    RANK() OVER w1 AS region_rank,
    RANK() OVER w2 AS continent_rank
FROM
    country
WHERE
    population > 1000000
WINDOW
    w1 AS (PARTITION BY region ORDER BY lifeexpectancy DESC),
    w2 AS (PARTITION BY continent ORDER BY lifeexpectancy DESC)
ORDER BY
    region_rank

   postgresql://postgres:***@localhost/imdb
   postgresql://postgres:***@localhost:5432/imdb
   postgresql://postgres:***@localhost:5432/mds
 * postgresql://postgres:***@localhost:5432/world
154 rows affected.


[('Cameroon', 1, 12),
 ('Czech Republic', 1, 18),
 ('Israel', 1, 4),
 ('Cuba', 1, 3),
 ('Senegal', 1, 7),
 ('Canada', 1, 1),
 ('Mauritius', 1, 3),
 ('Chile', 1, 1),
 ('Sri Lanka', 1, 10),
 ('South Africa', 1, 17),
 ('Libyan Arab Jamahiriya', 1, 1),
 ('Italy', 1, 3),
 ('United Kingdom', 1, 10),
 ('Papua New Guinea', 1, 3),
 ('Sweden', 1, 1),
 ('Estonia', 1, 29),
 ('Switzerland', 1, 1),
 ('Australia', 1, 1),
 ('Singapore', 1, 2),
 ('Japan', 1, 1),
 ('Costa Rica', 1, 4),
 ('Jordan', 2, 5),
 ('Hong Kong', 2, 3),
 ('Spain', 2, 4),
 ('Malaysia', 2, 16),
 ('Ireland', 2, 14),
 ('Ghana', 2, 8),
 ('Slovakia', 2, 20),
 ('Chad', 2, 21),
 ('Norway', 2, 6),
 ('Panama', 2, 6),
 ('Puerto Rico', 2, 5),
 ('Lesotho', 2, 19),
 ('Iran', 2, 18),
 ('Lithuania', 2, 30),
 ('Tunisia', 2, 2),
 ('Eritrea', 2, 10),
 ('Uruguay', 2, 2),
 ('New Zealand', 2, 2),
 ('France', 2, 4),
 ('United States', 2, 2),
 ('Tajikistan', 3, 29),
 ('Kuwait', 3, 7),
 ('Greece', 3, 7),
 ('Gabon', 3, 23),
 ('Namibia', 3, 38),
 ('Jamaica', 3, 7),
 ('Mexico', 3, 9),
 ('Algeria', 3, 4),
 ('Madagascar', 3, 11),
 ('Latvia', 3, 31),
 ('Netherlands', 3, 8),
 ('Togo', 3, 13),
 ('Taiwan', 3, 6),
 ('Poland', 3, 22),
 ('Vietnam', 3, 19),
 ('Finland', 3, 12),
 ('Argentina', 3, 3),
 ('South Korea', 4, 8),
 ('Dominican Republic', 4, 8),
 ('Thailand', 4, 20),
 ('Congo, The Democratic Republic of the', 4, 25),
 ('Honduras', 4, 10),
 ('Tanzania', 4, 15),
 ('Belgium', 4, 9),
 ('Gambia', 4, 14),
 ('Paraguay', 4, 4),
 ('Hungary', 4, 26),
 ('Uzbekistan', 4, 30),
 ('United Arab Emirates', 4, 9),
 ('Portugal', 4, 16),
 ('Swaziland', 4, 40),
 ('Denmark', 4, 15),
 ('Morocco', 4, 5),
 ('Oman', 5, 10),
 ('Kyrgyzstan', 5, 31),
 ('Bulgaria', 5, 27),
 ('Indonesia', 5, 22),
 ('China', 5, 12),
 ('Nigeria', 5, 16),
 ('Kenya', 5, 26),
 ('Austria', 5, 10),
 ('El Salvador', 5, 11),
 ('Venezuela', 5, 5),
 ('Egypt', 5, 6),
 ('Congo', 5, 27),
 ('Slovenia', 5, 17),
 ('Botswana', 5, 41),
 ('Trinidad and Tobago', 5, 13),
 ('Germany', 6, 12),
 ('Haiti', 6, 15),
 ('Central African Republic', 6, 36),
 ('Nicaragua', 6, 12),
 ('Burundi', 6, 30),
 ('Somalia', 6, 30),
 ('North Korea', 6, 17),
 ('Romania', 6, 28),
 ('Palestine', 6, 12),
 ('Sudan', 6, 9),
 ('Ecuador', 6, 6),
 ('Philippines', 6, 24),
 ('Macedonia', 6, 19),
 ('Kazakstan', 6, 32),
 ('Liberia', 6, 18),
 ('Lebanon', 7, 14),
 ('India', 7, 34),
 ('Cambodia', 7, 40),
 ('Mauritania', 7, 19),
 ('Angola', 7, 43),
 ('Mongolia', 7, 25),
 ('Croatia', 7, 20),
 ('Colombia', 7, 7),
 ('Belarus', 7, 32),
 ('Guatemala', 7, 14),
 ('Myanmar', 8, 41),
 ('Russian Federation', 8, 33),
 ('Turkey', 8, 15),
 ('Peru', 8, 8),
 ('Ethiopia', 8, 34),
 ('Benin', 8, 22),
 ('Pakistan', 8, 35),
 ('Yugoslavia', 8, 23),
 ('Laos', 9, 42),
 ('Bolivia', 9, 9),
 ('Syria', 9, 21),
 ('Albania', 9, 24),
 ('Ukraine', 9, 34),
 ('Turkmenistan', 9, 36),
 ('Guinea-Bissau', 9, 24),
 ('Uganda', 9, 37),
 ('Bangladesh', 10, 37),
 ('Bosnia and Herzegovina', 10, 25),
 ('Saudi Arabia', 10, 23),
 ('Burkina Faso', 10, 28),
 ('Mali', 10, 28),
 ('Rwanda', 10, 41),
 ('Brazil', 10, 10),
 ('Moldova', 10, 35),
 ('Zimbabwe', 11, 44),
 ('Nepal', 11, 39),
 ('Iraq', 11, 26),
 ('Malawi', 12, 45),
 ('Armenia', 12, 27),
 ('Guinea', 12, 32),
 ('Bhutan', 12, 43),
 ('Sierra Leone', 13, 33),
 ('Georgia', 13, 28),
 ('Afghanistan', 13, 44),
 ('Mozambique', 13, 46),
 ('Azerbaijan', 14, 33),
 ('Côte d\x92Ivoire', 14, 34),
 ('Zambia', 14, 47),
 ('Niger', 15, 39),
 ('Yemen', 15, 38)]

<br><br><br>

## Indexing and performance (optional)

Check the optional notebook if you want to learn more about indexing and performance.